core/http-auth/src/test/java/software/amazon/awssdk/http/auth/AwsV4HttpSignerTest.java

```diff
@@ -398,4 +377,79 @@ public void sign_withCrc32Checksum_and_withTrailinggHeader_shouldSignWithCrc32Ch
+
+        SyncSignRequest<? extends AwsCredentialsIdentity> request = generateBasicRequest(
+            AwsCredentialsIdentity.create("access", "secret"),
+            (httpRequest -> httpRequest.putHeader("x-amz-sha256", "required")),
+            (signRequest -> signRequest.putProperty(SIGNING_CLOCK, clock))
+        );
+
+        SyncSignedRequest signedRequest = signer.sign(request);
+
+        assertThat(signedRequest.request().firstMatchingHeader("Authorization"))
+            .hasValue(expectedAuthorizationHeaderWithSha256Header);
+    }
+
+    @Test
+    public void signAsync_withTickingClock_shouldUseSameTimeThroughoutSigning() {
+        final String expectedAuthorizationHeaderWithSha256Header =
+            AWS_4_HMAC_SHA_256_AUTHORIZATION +
+                "SignedHeaders=host;x-amz-archive-description;x-amz-date;x-amz-sha256, " +
+                "Signature=e73e20539446307a5dc71252dbd5b97e861f1d1267456abda3ebd8d57e519951";
+
+        // use a "ticking" clock that ticks each time an instant is requested - this lets us
+        // confirm that the same instant is used throughout the request
+        TickingClock clock = new TickingClock(Instant.ofEpochMilli(351153000968L));
+
+        AsyncSignRequest<? extends AwsCredentialsIdentity> request = generateBasicAsyncRequest(
+            AwsCredentialsIdentity.create("access", "secret"),
+            (httpRequest -> httpRequest.putHeader("x-amz-sha256", "required")),
+            (signRequest -> signRequest.putProperty(SIGNING_CLOCK, clock))
+        );
+
+        AsyncSignedRequest signedRequest = signer.signAsync(request);
+
+        assertThat(signedRequest.request().firstMatchingHeader("Authorization"))
+            .hasValue(expectedAuthorizationHeaderWithSha256Header);
+    }
+
+    @Test
+    public void sign_withoutRegionNameProperty_throws() {
+        SyncSignRequest<? extends AwsCredentialsIdentity> request = generateBasicRequest(
+            AwsCredentialsIdentity.create("access", "secret"),
+            (httpRequest -> {
+            }),
+            (signRequest -> signRequest.putProperty(SignerProperty.create(String.class, "RegionName"), null))
+        );
+
+        RuntimeException exception = assertThrows(RuntimeException.class, () -> signer.sign(request));
+
+        assertThat(exception.getMessage()).contains("must not be null");
+    }
+
+    @Test
+    public void sign_withoutServiceSigningNameProperty_throws() {
+        SyncSignRequest<? extends AwsCredentialsIdentity> request = generateBasicRequest(
+            AwsCredentialsIdentity.create("access", "secret"),
+            (httpRequest -> {
+            }),
+            (signRequest -> signRequest.putProperty(SignerProperty.create(String.class, "ServiceSigningName"), null))
+        );
+
+        RuntimeException exception = assertThrows(RuntimeException.class, () -> signer.sign(request));
+
+        assertThat(exception.getMessage()).contains("must not be null");
+    }
```

Comment:
s/NullPointerException/RuntimeException. Also add these tests for signAsync (more important).